File: android/Reguerta/app/src/main/java/com/reguerta/user/domain/startup/ResolveStartupVersionGateUseCase.kt
package com.reguerta.user.domain.startup

import com.reguerta.user.domain.RepositoryErrorKind
import com.reguerta.user.domain.RepositoryException
import java.net.URI

class ResolveStartupVersionGateUseCase(
    private val repository: StartupVersionPolicyRepository,
) {
    suspend operator fun invoke(
        platform: StartupPlatform,
        installedVersion: String,
    ): StartupVersionGateDecision {
        val policy = repository.getPolicy(platform)
        return evaluate(installedVersion = installedVersion, policy = policy)
    }

    fun evaluate(
        installedVersion: String,
        policy: StartupVersionPolicy,
    ): StartupVersionGateDecision {
        val comparisonToMinimum = SemanticVersionComparator.compare(installedVersion, policy.minimumVersion)
            ?: throw invalidVersionPolicy()
        val comparisonToCurrent = SemanticVersionComparator.compare(installedVersion, policy.currentVersion)
            ?: throw invalidVersionPolicy()
        val minimumToCurrent = SemanticVersionComparator.compare(policy.minimumVersion, policy.currentVersion)
            ?: throw invalidVersionPolicy()
        val storeUrl = policy.storeUrl.trim()

        if (minimumToCurrent > 0 || !storeUrl.isHttpUrl()) {
            throw invalidVersionPolicy()
        }

        return when {
            comparisonToMinimum < 0 && policy.forceUpdate ->
                StartupVersionGateDecision.ForcedUpdate(storeUrl = storeUrl)

            comparisonToMinimum < 0 || comparisonToCurrent < 0 ->
                StartupVersionGateDecision.OptionalUpdate(storeUrl = storeUrl)

            else -> StartupVersionGateDecision.Allow
        }
    }
}

private fun String.isHttpUrl(): Boolean = runCatching {
    val uri = URI(this)
    (uri.scheme.equals("http", ignoreCase = true) || uri.scheme.equals("https", ignoreCase = true)) &&
        !uri.host.isNullOrBlank()
}.getOrDefault(false)

private fun invalidVersionPolicy() = RepositoryException(
    kind = RepositoryErrorKind.INVALID_DATA,
    resource = "startupVersionPolicy",
)

File: ios/Reguerta/Reguerta/Domain/Startup/ResolveStartupVersionGateUseCase.swift
import Foundation

struct ResolveStartupVersionGateUseCase {
    private let storedRepository: any StartupVersionPolicyRepository
    private let environment: SessionEnvironment

    /// Loads the platform's version policy and resolves the startup gate.
    ///
    /// - Parameters:
    ///   - platform: The client platform whose remote policy should be loaded.
    ///   - installedVersion: The numeric, dot-separated application version.
    /// - Returns: The update or allow decision for the installed application.
    /// - Throws: A repository error or `RepositoryError.invalidData` when the policy cannot be
    ///   evaluated safely.
    func execute(platform: StartupPlatform, installedVersion: String) async throws -> StartupVersionGateDecision {
        let policy = try await storedRepository.policy(for: platform, environment: environment)
        return try evaluate(installedVersion: installedVersion, policy: policy)
    }

    /// Evaluates an installed application version against a validated startup policy.
    ///
    /// The policy is invalid when a version is malformed, the minimum exceeds the current
    /// version, or the store URL lacks an HTTP(S) scheme and host. Versions below the minimum
    /// are forced only when `forceUpdate` is enabled; all other older versions receive an
    /// optional update.
    ///
    /// - Parameters:
    ///   - installedVersion: The numeric, dot-separated application version.
    ///   - policy: The minimum, current, force-update, and store-link contract to apply.
    /// - Returns: `.forcedUpdate`, `.optionalUpdate`, or `.allow`.
    /// - Throws: `RepositoryError.invalidData` when the policy or installed version is invalid.
    func evaluate(installedVersion: String, policy: StartupVersionPolicy) throws -> StartupVersionGateDecision {
        guard let comparisonToMinimum = SemanticVersionComparator.compare(installedVersion, policy.minimumVersion),
              let comparisonToCurrent = SemanticVersionComparator.compare(installedVersion, policy.currentVersion),
              let minimumToCurrent = SemanticVersionComparator.compare(policy.minimumVersion, policy.currentVersion),
              minimumToCurrent <= 0,
              let storeURL = validatedStoreURL(policy.storeURL)
        else {
            throw RepositoryError.invalidData(resource: "startup.versionPolicy")
        }

        if comparisonToMinimum < 0 && policy.forceUpdate {
            return .forcedUpdate(storeURL: storeURL)
        }

        if comparisonToMinimum < 0 || comparisonToCurrent < 0 {
            return .optionalUpdate(storeURL: storeURL)
        }

        return .allow
    }

    private func validatedStoreURL(_ rawValue: String) -> String? {
        let normalized = rawValue.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let components = URLComponents(string: normalized),
              let scheme = components.scheme?.lowercased(),
              scheme == "https" || scheme == "http",
              components.host?.isEmpty == false
        else {
            return nil
        }
        return normalized
    }
}

extension ResolveStartupVersionGateUseCase {
    init(repository: any StartupVersionPolicyRepository, environment: SessionEnvironment) {
        self.storedRepository = repository
        self.environment = environment
    }
}
